freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaTopologyService.java

```diff
@@ -0,0 +1,163 @@
+            Integer remainingConnectionsForNode = entry.getValue();
+            for (int i = remainingConnectionsForNode; i > 0; i--) {
+                // Note: It is not possible for itr2 to reference on object with 0 remaining connections
+                // because itr2 is after the node iterator and the node iterator is checked for
+                // remaining connections.
+                if (itr2 == null || !itr2.hasNext()) {
+                    itr2 = getIteratorAfterCurrentNode(node, connectionRemaining.keySet());
+                    if (!itr2.hasNext()) {
+                        break;
+                    }
+                }
+                String node2 = itr2.next();
+                connectionRemaining.put(node2, connectionRemaining.get(node2) - 1);
+                ret.add(new UnorderedPair(node, node2));
+            }
+        }
+        return ret;
+    }
+
+    @SuppressWarnings("checkstyle:EmptyBlock")
+    private Iterator<String> getIteratorAfterCurrentNode(String node, Set<String> keySet) {
+        Iterator<String> ret = keySet.iterator();
+        while (ret.hasNext()) {
+            if (ret.next().equals(node)) {
+                break;
+            }
+        }
+        return ret;
+    }
+
+    private void createMissingSegments(FreeIpaClient freeIpaClient, String topologySuffixCn, Set<TopologySegment> topology) throws FreeIpaClientException {
+        Set<UnorderedPair> existingTopology = freeIpaClient.findTopologySegments(topologySuffixCn).stream()
+                .map(segment -> new UnorderedPair(segment.getLeftNode(), segment.getRightNode()))
+                .collect(Collectors.toSet());
+        Set<TopologySegment> segmentsToAdd = topology.stream()
+                .filter(segment -> !existingTopology.contains(new UnorderedPair(segment.getLeftNode(), segment.getRightNode())))
+                .collect(Collectors.toSet());
+        for (TopologySegment segment : segmentsToAdd) {
+            freeIpaClient.addTopologySegment(topologySuffixCn, segment);
+        }
+    }
+
+    private void removeExtraSegements(FreeIpaClient freeIpaClient, String topologySuffixCn, Set<TopologySegment> topology) throws FreeIpaClientException {
+        Set<UnorderedPair> topologyToKeep = topology.stream()
+                .map(segment -> new UnorderedPair(segment.getLeftNode(), segment.getRightNode()))
+                .collect(Collectors.toSet());
+        Set<TopologySegment> segmentsToRemove = freeIpaClient.findTopologySegments(topologySuffixCn).stream()
+                .filter(segment -> !topologyToKeep.contains(new UnorderedPair(segment.getLeftNode(), segment.getRightNode())))
+                .collect(Collectors.toSet());
+        for (TopologySegment segment : segmentsToRemove) {
+            freeIpaClient.deleteTopologySegment(topologySuffixCn, segment);
+        }
+    }
+
+    @VisibleForTesting
+    public static class UnorderedPair {
+        private final String l;
+
+        private final String r;
```

Comment:
could you rename these to `left` and `right` as their get method states also. thanks